### PR TITLE
[KED-2254] Add datatable csv dataset (#592)

### LINKED ISSUE #592: Integrate datable to Kedro extras datasets
## Description
Add [datatable](https://github.com/h2oai/datatable) to Kedro's extras datasets

## Context
[datatable](https://github.com/h2oai/datatable) is derived from R's [data.table](https://github.com/Rdatatable/data.table). As of 20/09/2020 ([Version 0.11.0](https://datatable.readthedocs.io/en/latest/releases/v0.11.0.html)), datatable is fully supported on Windows, Linux and MacOS and can be installed through pip. Offering datable in Kedro's extras dataset could offer many advantages in building pipelines:

- It would make the transition for R-users wanting to use Kedro a lot easier.
- Datable supports CSV, XLSX, and plain text as well as .tar, .gz, .zip, .gz2, and .tgz.
- Certain performance advantages over pandas (especially for datasets larger than 50GB): [benchmark](https://h2oai.github.io/db-benchmark/)
- fread is really fast at reading large CSVs. It also automatically detects separators.
- datatable's df.to_csv has an append parameter which appends the dataframe with column names if the file does not exist yet. If the file already exists, the column names are not added. Pandas to_csv offers "mode='a'" and "header=True" parameters which means choosing between rewriting the column names with each append or not having any column names. This is really practical for pipelines where you add new data from an IncrementalDatasets folder to a concatenated CSV.

## Possible Implementation
- Using the code from [pandas.CSVDataSet](https://github.com/quantumblacklabs/kedro/blob/master/kedro/extras/datasets/pandas/csv_dataset.py), we could change the "pd.read_csv(fs_file, **self._load_args)" to "dt.fread(fs_file, **self._load_args)". Saving the file should be similar code albeit different parameter names.
- For ExcelDataSets, we could also use code from [pandas.ExcelDataSet](https://github.com/quantumblacklabs/kedro/blob/master/kedro/extras/datasets/pandas/excel_dataset.py). Datatable also uses the xlrd engine for xlsx. Replace "pd.read_excel(fs_file, **self._load_args)" with "dt.fread(fs_file, **self._load_args)". Datatable does not output xlsx. So either we do not implement ExcelDataSets for datatable or we convert the datatable to pd.DataFrame and write to excel using pd.DataFrame.to_excel.

### PR DESCRIPTION
**Pull request by mlisovyi**
## Description
Adding dataset to read/write csv using [datatable](https://datatable.readthedocs.io/en/latest/index.html) instead of pandas.
Resolves #592 

## Development notes
I have added only CSV and not Excel as a Dataset. The later generates a lot of problems, when operating with `fsspec`. And when i try to use the file name directly I anyway fail the basic tests as for some reason write -> read sequence on Excel file does not yield the same data type of columns `datatable.Frame` as in the original `datatable.Frame`.

**To be discussed:**
What is the purpose of this add-on. While implementing it, i realized that:
* one gets a `datatable.Frame` object instead of `pandas.DataFrame`:
  * for R users, that are familiar with `data.table` in R this is beneficial as it provides a familiar API instead of pandas API
  * but it runs only on CSV as input
  * but it can be confusing that having CSVDataset but with a different prefix one gets an object that has a very different API
* one gets performance gain of datatable over pandas:
  * multi-thread reading of CSV
  * optimised string storage, i.e. smaller memory footprint by `object` column type
  * out-of-core computation (one can in principle process data that is larger than RAM)
  * but only for CSV input and maybe it would rather make sense to make a wrapper around `pandas.XXXDataset` to be able to read all pandas-supported formats and then benefit from datatable performance.

## Checklist

- [x] Read the [contributing](https://github.com/quantumblacklabs/kedro/blob/master/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](https://github.com/quantumblacklabs/kedro/blob/master/RELEASE.md) file
- [x] Added tests to cover my changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

**Review by lucasjamar — COMMENTED** [review]
I think you need to use datatable >=0.11.0 because it was not compatible to windows prior to this release: https://datatable.readthedocs.io/en/latest/releases/v0.11.0.html
Maybe this is what is causing the code to fail the windows pip compile and unit tests in CircleCI?

**Comment by mlisovyi** [comment]
@lucasjamar good point. Thanks for the suggestion!
We will need to have a look into datatable version. The `0.11.0` version on linux gave me a lot of headache :( When i pip-installed it into the nominal conda environment that had all kedro-requested packages included, the environment went into some weird broken state due to version conflicts and i was not able to do  do even `conda list`. Only uninstalling datatable with pip helped to get the environment into working state.
This requires more work to be invested to understand what goes wrong.

**Comment by lucasjamar** [comment]
> 
> 
> @lucasjamar good point. Thanks for the suggestion!
> We will need to have a look into datatable version. The `0.11.0` version on linux gave me a lot of headache :( When i pip-installed it into the nominal conda environment that had all kedro-requested packages included, the environment went into some weird broken state due to version conflicts and i was not able to do do even `conda list`. Only uninstalling datatable with pip helped to get the environment into working state.
> This requires more work to be invested to understand what goes wrong.

@mlisovyi  I already mentionned this [conda issue](https://github.com/h2oai/datatable/issues/2746) to datatable. Its a real bummer.

**Comment by mlisovyi** [comment]
The issue with datatable has been fixed on datatable side. But we need to wait for the next bugfix(?) release (supposedly 0.11.1)

**Comment by yetudada** [comment]
Thank you so much @mlisovyi! We'll get this reviewed!

File: kedro/extras/datasets/datatable/__init__.py
Comment (MerelTheisenQB) [inline kedro/extras/datasets/datatable/__init__.py:32]: I'd suggest using a different name to make it easier to distinguish from the pandas `CSVDataSet`. Perhaps `CSVDataTableDataSet`?

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (MerelTheisenQB) [inline kedro/extras/datasets/datatable/csv_dataset.py:134]: Any specific reason why you're using `r` mode here? We've recently updated a lot of our datasets to just use the default modes from fsspec, so if that works for this dataset as well I'd suggest removing this line.

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (MerelTheisenQB) [inline kedro/extras/datasets/datatable/csv_dataset.py:159]: Why are you still using `pandas` to save the data?

**Review by MerelTheisenQB — COMMENTED** [review]
Thanks for the contribution @mlisovyi ! I added a couple of questions about the implementation.

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py:159]: Indeed, that is inefficient and wierd.
The reason is that datatable doesn't seem to support file-like object as input for writing out a table. Therefore one can not benefit from access to multiple various filesystems that is provided by `fsspec.open()`

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py:134]: sure, i believe this will work dropping the line. I will try it out and take it down.

There was no good reason apriory- it was inherited from somewhere else, i believe :)

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (MerelTheisenQB) [reply kedro/extras/datasets/datatable/csv_dataset.py:159]: Right, so the main purpose of this dataset would be that you can load data in `datatable` and then work with it, before saving it again as pandas? If so, I'd suggest to update the description of the class to make this clear.

Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py:159]: yes, correct. the main 2 benefits would be:
* speed-up of csv reading, as `datatable` allows multi-thread reading from csv
* usage of data manipulation API that is familiar to R users, who might not have experience with `pandas`.
There is a nice list of advantages summarised in the original issue (#592).

Ok, i'll add doc-string of the class. This and the other proposed change will take a couple of days due to other tasks on my TODO.

Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py:166]: ok, the changed have been pushed

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py]: ok, the changed have been pushed

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (MerelTheisenQB) [inline kedro/extras/datasets/datatable/csv_dataset.py:58]: Thanks, this is a great explanation! 👏

**Review by MerelTheisenQB — APPROVED** [review]
This looks good to me! 👍  Can you please add your name to the `RELEASE.md` list of contributors and we will be good to go! 🎉

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (DmitriiDeriabinQB) [inline kedro/extras/datasets/datatable/csv_dataset.py:99]:
```suggestion
            save_args: Pandas options for saving CSV files.
```

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (DmitriiDeriabinQB) [reply kedro/extras/datasets/datatable/csv_dataset.py:166]: According to [`Frame.to_csv()`](https://datatable.readthedocs.io/en/latest/api/frame/to_csv.html) docs, "If no path is given, then the Frame will be serialized into a string, and that string will be returned".

So maybe we should just get such string, `.encode("utf-8")` it and send to fsspec instead of converting to pandas?

File: test_requirements.txt
Comment (DmitriiDeriabinQB) [inline test_requirements.txt:8]: Datatable should also be added into `extras_require` in `setup.py`

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py:166]: Unfortunately, I do not have enough experience to judge on it. I can blindly implement your suggestion, but I will not be able to validate it due to lack of time.

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py]: Resolving, as there has been no feedback since 6 days

File: kedro/extras/datasets/datatable/csv_dataset.py
Comment (mlisovyi) [reply kedro/extras/datasets/datatable/csv_dataset.py:166]: Unfortunately, I do not have enough experience to judge on it. I can blindly implement your suggestion, but I will not be able to validate it due to lack of time.

File: test_requirements.txt
Comment (mlisovyi) [reply test_requirements.txt:8]: Good catch! Thanks.

I'm not really familiar with the `extras_require` configuration for setuptools. I have added requirements as far as I get the logic, but could you please have a look that it is correct? In particular the cross-dependency of entries with datatable and pandas requirements is not fully clear to me.

**Comment by mlisovyi** [comment]
the failures in the py3.7 and py3.8 linting are very unclear to me. After a quick googling, my guess is that it is related to PyCQA/pylint#3318. I have no clue why does it does not appear in linting check with other python versions and my knowledge at this stage does not allow to pursue it further

**Comment by MerelTheisenQB** [comment]
> the failures in the py3.7 and py3.8 linting are very unclear to me. After a quick googling, my guess is that it is related to [PyCQA/pylint#3318](https://github.com/PyCQA/pylint/issues/3318). I have no clue why does it not appear in linting check with other python versions and my knowledge at this stage does not allow to pursue it further

We're seeing these errors on other builds as well, so they're not related to your PR. I'll investigate further.

**Comment by MerelTheisenQB** [comment]
Hi @mlisovyi , will you be able to finish this PR or are you happy with me jumping in and making it ready for merging?

**Comment by MerelTheisenQB** [comment]
Hi @mlisovyi, will you be able to address the comments on this PR? We're looking at cleaning up the open PR's, so we'll be deleting this in 2 weeks. Thanks!

**Comment by mlisovyi** [comment]
Hi @MerelTheisenQB ,

my apologies, unfortunately I'm currently overwhelmed with tasks at work and I do not see me finding time in the nearest future to contribute to the project. So I think this has to be closed to reduce maintenance burden for the package maintainers. If somewhere will come back to the original issue and will work on it one can always find the branch in my fork.

Once again my apologies.

**Comment by MerelTheisenQB** [comment]
Thanks for getting back about this @mlisovyi 🙂  No worries at all! Hopefully we'll see you back on the Kedro repo again in the future 🚀

**Pull request closed without merging**
